src/private_broadcast.cpp

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2023-present The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or https://opensource.org/license/mit/.
+
+#include <private_broadcast.h>
+#include <util/check.h>
+
+#include <algorithm>
+#include <tuple>
+
+/// If a transaction is not received back from the network for this duration
+/// after it is broadcast, then we consider it stale / for rebroadcasting.
+static constexpr auto STALE_DURATION{1min};
+
+bool PrivateBroadcast::Add(const CTransactionRef& tx) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+{
+    LOCK(m_mutex);
+    const bool inserted{m_transactions.emplace(tx, std::vector<SendStatus>{}).second};
+    return inserted;
+}
+
+std::optional<size_t> PrivateBroadcast::Remove(const CTransactionRef& tx)
+    EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+{
+    LOCK(m_mutex);
+    const auto handle{m_transactions.extract(tx)};
+    if (handle) {
+        const auto p{DerivePriority(handle.mapped())};
+        return p.num_confirmed;
+    }
+    return std::nullopt;
+}
+
+std::optional<CTransactionRef> PrivateBroadcast::PickTxForSend(const NodeId& will_send_to_nodeid)
+    EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+{
+    LOCK(m_mutex);
+
+    std::optional<CTransactionRef> most_urgent;
+    Priority most_urgent_prio;
+    decltype(m_transactions)::iterator most_urgent_it;
+
+    for (auto it = m_transactions.begin(); it != m_transactions.end(); ++it) {
+        const auto& [tx, sent_to]{*it};
+        const Priority tx_prio{DerivePriority(sent_to)};
+        if (!most_urgent.has_value() || most_urgent_prio < tx_prio) {
+            most_urgent = tx;
+            most_urgent_prio = tx_prio;
+            most_urgent_it = it;
+        }
+    }
+
+    if (most_urgent.has_value()) {
+        auto& sent_to{most_urgent_it->second};
+        sent_to.push_back(SendStatus{.nodeid = will_send_to_nodeid,
+                                     .picked = NodeClock::now(),
+                                     .confirmed = std::nullopt});
+    }
+
+    return most_urgent;
+}
+
+std::optional<CTransactionRef> PrivateBroadcast::GetTxForNode(const NodeId& nodeid)
+    EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+{
+    LOCK(m_mutex);
+    const auto tx_and_status{GetSendStatusByNode(nodeid)};
+    if (tx_and_status.has_value()) {
+        return tx_and_status.value().tx;
+    }
```

Comment:
I find this less repetitive (especially without needing `has_value()` and `.value()`) and that the scope of the variable is reduced - but I won't suggest it anymore if you dislike it:
```suggestion
    if (const auto tx_and_status{GetSendStatusByNode(nodeid)}) {
        return tx_and_status->tx;
    }
```

Reply:
Both seem to be valid. I prefer the current one.